integrations/azure-devops/azure_devops/client.py

```diff
@@ -0,0 +1,268 @@
+import json
+from typing import Any, AsyncGenerator, Optional
+from azure_devops.webhooks.webhook_event import WebhookEvent
+from port_ocean.context.event import event
+from port_ocean.context.ocean import ocean
+from loguru import logger
+from .http.http_client import HTTPBaseClient
+from azure_devops.utils import cache_results
+
+API_URL_PREFIX = "_apis"
+
+
+class AzureDevopsClient(HTTPBaseClient):
+    def __init__(self, organization_url: str, personal_access_token: str) -> None:
+        super().__init__(personal_access_token)
+        self._organization_base_url = organization_url
+
+    @classmethod
+    def create_from_ocean_config(cls) -> "AzureDevopsClient":
+        if cache := event.attributes.get("azure_devops_client"):
+            return cache
+        azure_devops_client = cls(
+            ocean.integration_config["organization_url"],
+            ocean.integration_config["personal_access_token"],
+        )
+        event.attributes["azure_devops_client"] = azure_devops_client
+        return azure_devops_client
+
+    @cache_results("projects")
+    async def generate_projects(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        params = {"includeCapabilities": "true"}
+        projects_url = f"{self._organization_base_url}/{API_URL_PREFIX}/projects"
+        async for projects in self._get_paginated_by_top_and_continuation_token(
+            projects_url, additional_params=params
+        ):
+            yield projects
+
+    @cache_results("teams")
+    async def generate_teams(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        teams_url = f"{self._organization_base_url}/{API_URL_PREFIX}/teams"
+        async for teams in self._get_paginated_by_top_and_skip(teams_url):
+            yield teams
+
+    async def generate_members(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        async for teams in self.generate_teams():
+            for team in teams:
+                members_in_teams_url = f"{self._organization_base_url}/{API_URL_PREFIX}/projects/{team['projectId']}/teams/{team['id']}/members"
+                async for members in self._get_paginated_by_top_and_skip(
+                    members_in_teams_url
+                ):
+                    for member in members:
+                        member["teamId"] = team["id"]
+                    yield members
+
+    @cache_results("repositories")
+    async def generate_repositories(self) -> AsyncGenerator[list[dict[Any, Any]], None]:
+        async for projects in self.generate_projects():
+            for project in projects:
+                repos_url = f"{self._organization_base_url}/{project['id']}/{API_URL_PREFIX}/git/repositories"
+                repositories = self._parse_response_values(
+                    await self.send_request("GET", repos_url)
+                )
+                yield repositories
```

Comment:
each query will be execute synchronically whe you use for and await

Reply:
There is a way to return repositories not by project, But the response coming from azure devops isn't paginatable (https://learn.microsoft.com/en-us/rest/api/azure/devops/git/repositories/list?view=azure-devops-rest-7.1&tabs=HTTP)
Should I return it as returned from the api without being able to paginate this?